source/agora/consensus/BlockRewardCalc.d

```diff
@@ -0,0 +1,651 @@
+
+        // extra reward for the the last block of the period
+        // block reward payout spans adjecent periods
+        actual_rewards = getTestRewards!"TotalValidatorReward"(Height(0), Height(19), 2, 3,
+            [BlockRewardsTup(Amount(16), 5), BlockRewardsTup(Amount(26), 5)]);
+        assert(getAmountArr([0, 0, 0, 0, 0, 9, 0, 0, 12, 0, 0, 15, 0, 0, 6, 0, 0, 0, 0]) == actual_rewards);
+
+        // extra reward for the the last block of the period
+        // block reward payout spans adjecent periods
+        // periods have different length
+        actual_rewards = getTestRewards!"TotalValidatorReward"(Height(0), Height(22), 2, 3,
+            [BlockRewardsTup(Amount(17), 4), BlockRewardsTup(Amount(7), 3), BlockRewardsTup(Amount(15), 6)]);
+        assert(getAmountArr([0, 0, 0, 0, 0, 12, 0, 0, 9, 0, 0, 7, 0, 0, 6, 0, 0, 5, 0, 0, 0, 0]) == actual_rewards);
+
+        // Checking whether we conform to the whitepaper's
+        // coin generation rate of 27 BOA coins per 5 second for the first year.
+        actual_rewards = getTestRewards!"TotalValidatorReward"(Height(0), Height(22), 2, 3,
+        ConsensusConfig().validator_block_rewards, 100, 5.seconds);
+        assert(getAmountArr([0, 0, 0, 0, 0, 810000000, 0, 0, 810000000, 0, 0, 810000000, 0, 0,
+                             810000000, 0, 0, 810000000, 0, 0, 810000000, 0]) == actual_rewards);
+    }
+
+    /***************************************************************************
+
+    Returns the total validator reward after applying the penalty for
+    missing signatures.
+
+    Params:
+        height = the height of the block that will contain the CoinBase transaction
+        comply_perc = (#actual signatures / #expected signatures) * 100 on all
+                      the blocks for which we want to calculate the block reward
+        indent = number of spaces before log messages
+
+    Returns:
+        validator whitepaper reward - penalty on validators due to missing signature
+
+    ***************************************************************************/
+
+    public Amount getReducedValidatorReward (in Height height, double comply_perc, uint indent = 0) @safe const nothrow
+    {
+        auto indentStr = ' '.repeat(indent).array;
+        log.trace("{}Calculating reduced validator reward", indentStr);
+        Amount total_validator_reward = getTotalValidatorReward(height, indent + indentInc);
+        ubyte reduce_perc = cast(ubyte) round(
+            consensus_params.BlockRewardFactorB *
+            exp(comply_perc * consensus_params.BlockRewardFactorA) +
+            consensus_params.BlockRewardFactorC);
+        log.trace("{}comply percentage: {}", indentStr, comply_perc);
+        log.trace("{}reduce percentage: {}", indentStr, reduce_perc);
+        reduce_perc = cast(ubyte) min(reduce_perc, 100);
+
+        if (reduce_perc == 100)
+            return total_validator_reward;
+        total_validator_reward.div(100);
```

Comment:
What about the remainder ?

Reply:
remainder goes to the Commons Budget and calculated by: getExtraCommonsBudgetReward